File: django_formwork/templatetags/formwork.py
from django import template
from django.templatetags.static import static
from django.utils.html import format_html

register = template.Library()


@register.simple_tag
def formwork_css() -> str:
    """Output a ``<link>`` tag for the formwork bridge CSS.

    Usage::

        {% load formwork %}
        {% formwork_css %}
    """
    return format_html('<link rel="stylesheet" href="{}">', static("formwork/formwork.css"))


@register.simple_tag
def formwork_js() -> str:
    """Output a ``<script type="module">`` tag for the formwork bundle.

    Loads ``formwork.js`` as an ES module, which imports the page-global
    core (htmx morph extension, dirty-tracking, native-validation
    disabling) and each widget's Alpine.data component from
    ``formwork/widgets/*.js``.

    For per-form loading via ``{{ form.media }}``, use
    ``{% formwork_core_js %}`` instead, which loads only the page-global
    core; widget JS is then included automatically per form via
    Django's Media plumbing.

    Usage::

        {% load formwork %}
        {% formwork_js %}
    """
    return format_html('<script type="module" src="{}"></script>', static("formwork/formwork.js"))


@register.simple_tag
def formwork_core_js() -> str:
    """Output a ``<script type="module">`` tag for the formwork core only.

    Loads ``formwork-core.js``: the page-global htmx morph extension,
    dirty-tracking, and native-validation disabling.  Use together with
    ``{{ form.media }}`` to load per-widget Alpine code on a per-form
    basis instead of via the full bundle.

    ``{% formwork_js %}`` already imports the core, so combining both
    tags is safe but redundant (ES module dedup by URL means the core
    only executes once).

    Usage::

        {% load formwork %}
        {% formwork_core_js %}
        {{ form.media }}
    """
    return format_html('<script type="module" src="{}"></script>', static("formwork/formwork-core.js"))


@register.filter
def attr(attrs: dict, key: str) -> str:
    """Look up a widget attr by its literal (possibly hyphenated) key.

    DTL cannot traverse hyphenated keys via dot syntax
    (``widget.attrs.aria-invalid`` is a syntax error) and has no
    subscript form, so this filter bridges the gap for keys like
    ``aria-invalid`` / ``aria-describedby`` that Django writes onto a
    widget's attrs.  Jinja2 templates use native subscription instead.
    """
    return attrs.get(key, "")

File: django_formwork/static/formwork/formwork.js
/**
 * formwork.js — formwork bundle.
 *
 * Aggregator that imports the page-global core (morph, dirty-tracking,
 * native-validation disabling) plus each widget's Alpine.data
 * component module.  Load via {% formwork_js %} or import directly
 * from a JS bundler entry (vite, esbuild, webpack, etc.).
 *
 * Three loading paths:
 *
 *   1. {% formwork_js %} — this file as <script type="module">.
 *      One request, everything loaded.
 *
 *   2. {% formwork_core_js %} + {{ form.media }} — page-global core
 *      from formwork-core.js, per-widget JS via Django's Media class.
 *      Two tags but only the JS the form actually uses is loaded.
 *
 *   3. Bundler import — `import "django-formwork/formwork.js"` from a
 *      JS bundler entry.  The bundler resolves the chain and emits a
 *      single bundle.  No {% formwork_js %} / {{ form.media }} needed.
 *
 * ES module URL deduplication makes the paths safely composable: if you
 * accidentally load both `{% formwork_js %}` and `{% formwork_core_js %}`,
 * the core only executes once.
 *
 * Prerequisites (loaded by the user, BEFORE this script):
 *   - htmx 4.x
 *   - Alpine.js 3.x (if using Alpine-powered widgets)
 */

import "./formwork-core.js";
import "./widgets/search_select.js";
import "./widgets/multi_select.js";
import "./widgets/combo_box.js";
import "./widgets/date_picker.js";
import "./widgets/drop_zone.js";
import "./widgets/image_upload.js";
import "./widgets/input_mask.js";
import "./widgets/input_number.js";
import "./widgets/otp_input.js";
import "./widgets/password_reveal.js";
import "./widgets/validated_textarea.js";
